rskj-core/src/main/java/co/rsk/peg/union/UnionBridgeSupportImpl.java

```diff
@@ -0,0 +1,111 @@
+
+    private final ActivationConfig.ForBlock activations;
+    private final UnionBridgeConstants constants;
+    private final UnionBridgeStorageProvider storageProvider;
+    private final SignatureCache signatureCache;
+
+    public UnionBridgeSupportImpl(ForBlock activations, UnionBridgeConstants constants, UnionBridgeStorageProvider storageProvider,
+        SignatureCache signatureCache) {
+        this.activations = activations;
+        this.constants = constants;
+        this.storageProvider = storageProvider;
+        this.signatureCache = signatureCache;
+    }
+
+    @Override
+    public int setUnionBridgeContractAddressForTestnet(@Nonnull Transaction tx, RskAddress unionBridgeContractAddress) {
+        final String SET_UNION_BRIDGE_ADDRESS_TAG = "setUnionBridgeContractAddressForTestnet";
+
+        logger.info("[{}] Setting new union bridge contract address: {}", SET_UNION_BRIDGE_ADDRESS_TAG, unionBridgeContractAddress);
+
+        // Check if the network is MAINNET as the contract address can only be set in testnet or regtest
+        if (isEnvironmentDisable()) {
+            return UnionResponseCode.ENVIRONMENT_DISABLED.getCode();
+        }
+
+        if (isUnauthorizedCaller(tx)) {
+            return UnionResponseCode.UNAUTHORIZED_CALLER.getCode();
+        }
+
+        if (isInvalidAddress(unionBridgeContractAddress)) {
+            return UnionResponseCode.INVALID_VALUE.getCode();
+        }
+
+        RskAddress currentUnionBridgeAddress = storageProvider.getAddress(activations).orElse(constants.getAddress());
+        if (isAddressAlreadyStored(currentUnionBridgeAddress, unionBridgeContractAddress)) {
+            return UnionResponseCode.INVALID_VALUE.getCode();
+        }
+
+        storageProvider.setAddress(unionBridgeContractAddress);
+        logger.info("[{}] Union Bridge Contract Address has been updated. Previous address: {} New address: {}", SET_UNION_BRIDGE_ADDRESS_TAG, currentUnionBridgeAddress, unionBridgeContractAddress);
+        return UnionResponseCode.SUCCESS.getCode();
+    }
+
+    private boolean isEnvironmentDisable() {
+        String currentNetworkId = constants.getBtcParams().getId();
+
+        boolean isEnvironmentDisable = currentNetworkId.equals(NetworkParameters.ID_MAINNET);
+        if (isEnvironmentDisable) {
+            String baseMessage = String.format("Union Bridge Contract Address can only be set in Testnet and RegTest environments. Current network: %s", currentNetworkId);
+            logger.warn(LOG_PATTERN, "isEnvironmentDisable", baseMessage);
+        }
+        return isEnvironmentDisable;
+    }
+
+    private boolean isUnauthorizedCaller(Transaction tx) {
+        // Check if the caller is isUnauthorizedCaller to set a new bridge contract address
+        AddressBasedAuthorizer authorizer = constants.getChangeAuthorizer();
+        boolean isUnauthorizedCaller = !authorizer.isAuthorized(tx, signatureCache);
+        if (!isUnauthorizedCaller) {
```

Comment:
carefulll

Reply:
Good catch. Logs were not being logged because of that.